server/pulp/server/db/querysets.py

```diff
@@ -84,3 +90,26 @@ def update_one(self, *args, **kwargs):
             raise NotImplementedError("The 'multi' parameter cannot be set on this method.")
         kwargs['multi'] = False
         self.update(*args, **kwargs)
+
+
+class RepoQuerySet(CriteriaQuerySet):
+    """
+    Custom queryset for repositories.
+    """
+
+    def get_repo_or_missing_resource(self, repo_id):
```

Comment:
The use of `MissingResource` is very inconsistent over the code base, but it does take args and kwargs which means that we can use it this way: `MissingResource(repository=repo_id)` 

If we were to catch `DoesNotExist` in the middleware, we would be notified that the Repository doesn't exist, but without reference to the data that the user expected to work. I am not necessarily opposed to doing it that way since it seems like the user *should* know what data they just sent. Also, it would be simpler and more consistent around the API, but it is in the gray area of backwards compatibility. 

If we were to do it that way, this is how our use would change:

``` python
>>> # Current/my implementation
>>> model.Repository.objects.get_repo_or_missing_resource('fake_repo')
Missing resource(s): repository=fake_repo

>>> # If the middleware were to catch DoesNotExist
>>> model.Repository.objects.get(repo_id='fake_repo')
DoesNotExist: Repository matching query does not exist.
```
To me, the main problem with backwards compatibility is that the MissingResource exception has a field `MissingResource.resources` which is directly returned via the API:

``` json
{
    "_href": "/pulp/api/v2/repositories/fake/", 
    "error": {
        "code": "PLP0009", 
        "data": {
            "resources": {
                "repository": "fake"
            }
        }, 
        "description": "Missing resource(s): repository=fake", 
        "sub_errors": []
    }, 
    "error_message": "Missing resource(s): repository=fake", 
    "exception": null, 
    "http_request_method": "GET", 
    "http_status": 404, 
    "resources": {
        "repository": "fake"
    }, 
    "traceback": null
}
```